### PR TITLE
PHOENIX-6506 : Tenant Connection is not able to access/validate Global Sequences

### PR DESCRIPTION
**Pull request by lokiore**

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  23m 55s |  master passed  |
| +0 |  hbaserecompile  |  30m 55s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  4s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m  8s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 53s |  master passed  |
| +0 :ok: |  spotbugs  |   3m  6s |  phoenix-core in master has 965 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  13m 39s |  the patch passed  |
| +0 |  hbaserecompile  |  23m 44s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  4s |  the patch passed  |
| -1 :x: |  checkstyle  |   1m 29s |  phoenix-core: The patch generated 14 new + 2227 unchanged - 0 fixed = 2241 total (was 2227)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 16s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 113m 14s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 43s |  The patch does not generate ASF License warnings.  |
|  |   | 182m  6s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1261 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux 416eee8a37f0 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 62f63c8 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/1/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/1/testReport/ |
| Max. process+thread count | 15650 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by mnpoonia** [comment]
@lokiore - Looks like there are checkstyle issues and some unit test are failing

File: phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java:5100]: nit: replace with stream -> filter -> findfirst?
```
return sequenceMap.entrySet().stream()
        .filter(entry -> compareSequenceKeysWithoutTenant(key, entry.getKey()))
        .findFirst()
        .map(Entry::getValue)
        .orElse(null);
```

File: phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java
Comment (virajjasani) [inline phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java:5111]: Here we are checking if key2 is global sequence right?

It's bit difficult to understand in one statement. Good to break it down, something like
```
    private boolean compareSequenceKeysWithoutTenant(SequenceKey keyToCompare,
            SequenceKey availableKey) {
        boolean sameSchema = keyToCompare.getSchemaName() == null
                ? availableKey.getSchemaName() == null
                : keyToCompare.getSchemaName().equals(availableKey.getSchemaName());
        if (!sameSchema) {
            return false;
        }
        if (availableKey.getTenantId() != null) {
            return false;
        }
        return keyToCompare.getSequenceName().equals(availableKey.getSequenceName());
    }
```

**Review by virajjasani — COMMENTED** [review]
Skimmed, left few comments

**Comment by virajjasani** [comment]
@abhishek-chouhan @dbwong Could you please take a look when you have some bandwidth?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:711]: I think we can just fail here with `Assert.fail()`

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:736]: same as above, just use `fail()`

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:50]: Please revert this to individual imports, I think you just need one additional import.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:688]: Nice coverage

**Review by virajjasani — COMMENTED** [review]
Left few more comments

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:707]: Adding `fail()` here would mean we will make sure above Upsert never succeeds.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/UpsertValuesIT.java:732]: same as above, good to add `fail()` here

**Comment by virajjasani** [comment]
@stoty I agree and we have 4.x backport PR #1270 with proper Java 7 semantics so I think we are good with this PR using streams and eventually land on master and 5.1.
4.x branch is still stuck with Java 7 because of HBase 1 support but our CI builds still use JDK 1.8 so it doesn't catch such issues.

**Review by virajjasani — APPROVED** [review]
+1, pending QA

**Comment by abhishek-chouhan** [comment]
I'm not sure if this is a bug. Tenants are allowed to create their own sequences and use those for their operations. In general tenant connections are not able to see data associated with a global non-tenant connection. This could also be considered similar to that. I do think there is value in having a shared sequence across tenants, however need to think more if that should be through enabling it generally, like this pr aims to do, or having something like a "multi-tenant" sequence (create sequnce .. multi-tenant=true). @gjacoby126 @virajjasani

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  22m 24s |  master passed  |
| +0 |  hbaserecompile  |  29m 16s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  3s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 26s |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 53s |  master passed  |
| +0 :ok: |  spotbugs  |   3m 22s |  phoenix-core in master has 964 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  14m 23s |  the patch passed  |
| +0 |  hbaserecompile  |  24m 34s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m  7s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  7s |  the patch passed  |
| -1 :x: |  checkstyle  |   1m 24s |  phoenix-core: The patch generated 20 new + 2221 unchanged - 6 fixed = 2241 total (was 2227)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 32s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 113m 34s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 18s |  The patch does not generate ASF License warnings.  |
|  |   | 179m 51s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1261 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaserebuild hbaseanti checkstyle compile |
| uname | Linux 69278de4d602 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 6842c1d |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/2/testReport/ |
| Max. process+thread count | 14213 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1261/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by virajjasani** [comment]
> I do think there is value in having a shared sequence across tenants, however need to think more if that should be through enabling it generally, like this pr aims to do, or having something like a "multi-tenant" sequence (create sequnce .. multi-tenant=true)

Nice one, yeah if not the global sequence, maybe a special type of multi-tenant sequence (i.e new attribute attached to sequence) that can make it accessible to all tenants.

**Comment by abhishek-chouhan** [comment]
Lets repurpose/open a new jira, which introduces multi-tenant sequences unless there are objections.

**Comment by gjacoby126** [comment]
@abhishek-chouhan @virajjasani - I believe the usual pattern in Phoenix is that a tenant connection should already be able to see global things. For example, if you get a tenant connection and query a non-MT table, you should be able to get rows back.  Tenant connections afaik just add "WHERE [TenantId Column] = [Your connection's tenantId]" to all multi-tenant tables.

**Comment by lokiore** [comment]
@gjacoby126 @abhishek-chouhan @virajjasani - What you guys think should be next steps for this!?
Is this good or we need to repurpose!?

**Comment by virajjasani** [comment]
> I believe the usual pattern in Phoenix is that a tenant connection should already be able to see global things. For example, if you get a tenant connection and query a non-MT table, you should be able to get rows back.

Tenant connection can read from global views as well. I think this approach also would be good enough i.e having global sequence accessible to tenant connection. If we are good to go with this approach, changes look good to me per my previous reviews. I was just curious about the final decision on the approach we decide to go ahead with.
@abhishek-chouhan WDYT? Does this sound good to you?

**Comment by abhishek-chouhan** [comment]
Makes sense. Should be fine to go ahead with this since it gives us parity with the current behavior as @gjacoby126 rightly pointed out.

**Comment by virajjasani** [comment]
Sounds good. Thanks @abhishek-chouhan @gjacoby126 for your valuable inputs.
Will merge this after 24 hr unless there are any objections on this PR.

**Pull request merged**
